### ISSUE TITLE
No adaptor selected when querying '/'?

### ISSUE THREAD

**Issue by mbroedl** [jspricke/radicale-storage#2]
Thought this is a great project to integrate my todo list with my Evolution PIM (Gnome).

For some reason, it throws an error at me when I try to load it in evolution. It seems as if there is no adapter passed through, but I've no idea how to fix it.

I have installed python 3.6, radicale 2.1.8, and the most recent version of your plugin.

Below the error message from `radicale --debug`, and the config file.
```
[7f1c7ddbb540] INFO: Starting Radicale
[7f1c7ddbb540] INFO: Authentication type is 'none'
[7f1c7ddbb540] INFO: Storage type is 'radicale_remind'
[7f1c7ddbb540] INFO: Rights type is 'None'
[7f1c7ddbb540] INFO: Web type is 'none'
[7f1c7ddbb540] INFO: Listening to 'myservername' on port 5232
[7f1c7ddbb540] INFO: Radicale server ready
[7f1c777f2700] INFO: REPORT request for '/' with depth '1' received from 127.0.0.1 using 'Evolution/3.26.4'
[7f1c777f2700] DEBUG: Request headers:
{'CONTENT_LENGTH': '328',
 'CONTENT_TYPE': 'application/xml; charset="utf-8"',
 'GATEWAY_INTERFACE': 'CGI/1.1',
 'HTTP_ACCEPT_LANGUAGE': 'en-gb, en;q=0.9, en;q=0.8',
 'HTTP_CACHE_CONTROL': 'no-cache',
 'HTTP_CONNECTION': 'close',
 'HTTP_DEPTH': '1',
 'HTTP_HOST': '0.0.0.0:5232',
 'HTTP_PRAGMA': 'no-cache',
 'HTTP_USER_AGENT': 'Evolution/3.26.4',
 'PATH_INFO': '/',
 'QUERY_STRING': '',
 'REMOTE_ADDR': '127.0.0.1',
 'REMOTE_HOST': '',
 'REQUEST_METHOD': 'REPORT',
 'SCRIPT_NAME': '',
 'SERVER_NAME': 'myservername',
 'SERVER_PORT': '5232',
 'SERVER_PROTOCOL': 'HTTP/1.1',
 'SERVER_SOFTWARE': 'WSGIServer/0.2',
 'wsgi.errors': <_io.StringIO object at 0x7f1c77d5dd38>,
 'wsgi.file_wrapper': <class 'wsgiref.util.FileWrapper'>,
 'wsgi.input': <_io.BufferedReader name=6>,
 'wsgi.multiprocess': False,
 'wsgi.multithread': True,
 'wsgi.run_once': False,
 'wsgi.url_scheme': 'http',
 'wsgi.version': (1, 0)}
[7f1c777f2700] DEBUG: Sanitized script name: ''
[7f1c777f2700] DEBUG: Sanitized path: '/'
[7f1c777f2700] DEBUG: Request content:
<?xml version="1.0"?>
<C:calendar-query xmlns="DAV:" xmlns:C="urn:ietf:params:xml:ns:caldav">
  <prop>
    <getetag />
  </prop>
  <C:filter>
    <C:comp-filter name="VCALENDAR">
      <C:comp-filter name="VTODO">
        <C:time-range end="20180315T151438Z" start="20180104T151438Z" />
      </C:comp-filter>
    </C:comp-filter>
  </C:filter>
</C:calendar-query>

[7f1c777f2700] ERROR: An exception occurred during REPORT request on '/': 'NoneType' object has no attribute 'get_uids'
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 321, in __call__
    status, headers, answers = self._handle_request(environ)
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 494, in _handle_request
    environ, base_prefix, path, user)
  File "/usr/lib/python3.6/site-packages/radicale/__init__.py", line 944, in do_REPORT
    base_prefix, path, xml_content, collection)
  File "/usr/lib/python3.6/site-packages/radicale/xmlutils.py", line 1250, in report
    multistatus):
  File "/usr/lib/python3.6/site-packages/radicale/xmlutils.py", line 1224, in retrieve_items
    yield from collection.get_all_filtered(filters)
  File "/usr/lib/python3.6/site-packages/radicale/storage.py", line 586, in <genexpr>
    return ((item, False) for item in self.get_all())
  File "/usr/lib/python3.6/site-packages/radicale_remind.py", line 114, in list
    for uid in self.adapter.get_uids(self.filename):
AttributeError: 'NoneType' object has no attribute 'get_uids'
```

```
[server]
hosts = 0.0.0.0:5232
ssl = False

[storage]
type = radicale_remind
filesystem_folder = /home
task_folder = /home/——/.task

[web]
type = none
```

**Comment by jspricke** [comment]
Thanks for writing. I was able to reproduce your bug added a warning to 
the adapter. But the actual problem is the URL you provided to 
Evolution. It must be of the form: 

http://myservername/username/.task/Project

With `Project` being the project in Taskwarrior.

I will keep this open for you to report your findings.

Cheers Jochen

**Comment by mbroedl** [comment]
Ah, thanks a lot, Jochen!

Had a few more error messages because some of my task did not have projects assigned, so I just added `'project' in task` into the if conditions in lines 89, 214, 224.

Editing etc work as well. However, evolution gives me an error message when saving an item or creating a new one.
Looking into the debug log, it seems as if evolution requests the item with the same ID, and since tags change the ID is not permanent, throwing an warning both in evolution and radicale; not essentially a problem, but it requires me to refresh the calendar to see the items in evolution. For saving a new one, it seems as if evolution makes up a hash like this `4ab00f464ef8fd871c5ed6f81f69b542d0074051.ics` and then wants it to stay that way.
```
[7f639a1e1700] INFO: PUT response status for '/malte/.task/home/4ae58b66-2144-4f70-80cc-d2f17becf970.cf4cdde8a46eb70f827857a9d8243eb31b0839b2@balte' in 0.036 seconds: 201 Created
[7f639a1e1700] INFO: GET request for '/malte/.task/home/4ae58b66-2144-4f70-80cc-d2f17becf970.cf4cdde8a46eb70f827857a9d8243eb31b0839b2@balte' received from 127.0.0.1 using 'Evolution/3.26.4'
...
[7f639a1e1700] DEBUG: Sanitized path: '/malte/.task/home/4ae58b66-2144-4f70-80cc-d2f17becf970.cf4cdde8a46eb70f827857a9d8243eb31b0839b2@balte'
[7f639a1e1700] DEBUG: Response content:
The requested resource could not be found.
[7f639a1e1700] INFO: GET response status for '/malte/.task/home/4ae58b66-2144-4f70-80cc-d2f17becf970.cf4cdde8a46eb70f827857a9d8243eb31b0839b2@balte' in 0.001 seconds: 404 Not Found
```

Neither is a problem really, but you asked for findings. ;)

**Comment by jspricke** [comment]
>Had a few more error messages because some of my task did not have projects assigned, so I just added `'project' in task` into the if conditions in lines 89, 214, 224.

Good point, I have `default.project=Default` in my `taskrc`. Would it 
help you to have tasks without a project exported with a `Default`?

Speaking of it, I'm really interested to hear from others how they use 
Taskwarrior and how this project would fit best for them. Could you 
comment if exporting projects as different lists make sense to you, or 
would it be better to have one URL for all tasks and expose the projects 
as `CATEGORIES` for example?

>Editing etc work as well. However, evolution gives me an error message when saving an item or creating a new one.

Thanks, I'm aware of this. Basically Evolution and Taskwarrior both 
assign an UID to the task. For tasks created in Taskwarrior that's not a 
problem, as the UID get's exported and is accepted by Evolution. But 
there is no easy way to import a Evolution UID into Taskwarrior. The 
only option would be to add it as a user definied attribute (UDA), but 
it feels a little hacky and would require more thought for the Remind 
and Abook parts. Maybe we can add a flag to allow this optionally, but 
it would be quiet low on my priority list.

Are you sure this happens upon saving an item (initially created in 
Taskwarrior) as well?

**Comment by mbroedl** [comment]
I actually just switched from other todo-managers, so I'm new to taskwarrior and still contemplating whether todo.txt suits me better because I like working in plain text even more than on CLI. ;)
So, I just put an '_all' project filename (as in for CalDav) in order to access all items at once regardless of project.

> Could you comment if exporting projects as different lists make sense to you, or would it be better to have one URL for all tasks and expose the projects as `CATEGORIES` for example?

Difficult question. I like having the projects colour separated, which would happen if I subscribe to them manually; the problem then is that one needs to subscribe to new/temporary projects separately. Maybe in the ideal world, one could subscribe to major projects separately, and the rest would be in a "remaining" list. But that's hard to know which ones these are unless manually excluded...

> But there is no easy way to import a Evolution UID into Taskwarrior.

There is no actual need to keep the ids permanent for evolution, as with every calendar update it accepts the taskwarrior ids; the only thing that's needed though is that within a session, evolution can still access the old/evolution-original id when requesting to know that the changes happened.
The only problem is that radicale seems to call the adapter every time as opposed to keeping it stable, but maybe there is still a way to store "old ids" in cache for a while and link them to the new entry?
But I agree, its not high priority.

>  Are you sure this happens upon saving an item (initially created in Taskwarrior) as well?

Unfortunately, yes.
I will spare you the most of the log.
```
[7f0a7c7d2700] INFO: PUT request for '/malte/.task/home/aa2a7640-04eb-45a4-9d48-d2577cbeb14b.21205197a9fe0bbdc8978425c73105479134b980@balte' received from 127.0.0.1 using 'Evolution/3.26.4'
...
[7f0a7c7d2700] INFO: PUT response status for '/malte/.task/home/aa2a7640-04eb-45a4-9d48-d2577cbeb14b.21205197a9fe0bbdc8978425c73105479134b980@balte' in 0.025 seconds: 201 Created
```
So the creation/changing of items works fine and is accepted. But then evolution asks for the same ical ID again to update itself:
```
[7f0a7c7d2700] INFO: GET request for '/malte/.task/home/aa2a7640-04eb-45a4-9d48-d2577cbeb14b.21205197a9fe0bbdc8978425c73105479134b980@balte' received from 127.0.0.1 using 'Evolution/3.26.4'
...
[7f0a7c7d2700] DEBUG: Response content:
The requested resource could not be found.
[7f0a7c7d2700] INFO: GET response status for '/malte/.task/home/aa2a7640-04eb-45a4-9d48-d2577cbeb14b.21205197a9fe0bbdc8978425c73105479134b980@balte' in 0.003 seconds: 404 Not Found
```
And many more failed requests.
When I reload the calendar, then the id for that same entry changed to:
`aa2a7640-04eb-45a4-9d48-d2577cbeb14b.3600a17805363ceabf970c4321250635ef3bed82@balte`
So the first part of the id (before the dot; which is used inside the script) remains, as you said, because it's internal to taskwarrior. The second part seems to be generated as a hash on the basis of the json-object if I see that correctly. So that part changes, and evolution cannot access it any more after the update. — Once I refresh the calendar, evolution gets the changes as well. Maybe one could force reload via a server message?

**Comment by jspricke** [comment]
* mbroedl <notifications@github.com> [2018-02-08 10:07]:
>So, I just put an '_all' project filename (as in for CalDav) in order to access all items at once regardless of project.

Would be great if you could send a pull request to get this integrated.

>Difficult question. I like having the projects colour separated, which would happen if I subscribe to them manually; the problem then is that one needs to subscribe to new/temporary projects separately. Maybe in the ideal world, one could subscribe to major projects separately, and the rest would be in a "remaining" list. But that's hard to know which ones these are unless manually excluded...

I fear so as well. Maybe we could have a list in the adapter which 
projects to export individually.

>So the first part of the id (before the dot; which is used inside the script) remains, as you said, because it's internal to taskwarrior. The second part seems to be generated as a hash on the basis of the json-object if I see that correctly. So that part changes, and evolution cannot access it any more after the update.

You are absolutely correct, sorry forgot about that part. I think I 
modeled it that way to make sure changes in Taskwarrior are pushed to 
the DAV side as well and to be sure to not get update conflicts.
Maybe putting the second part into the etag would be better.

>Once I refresh the calendar, evolution gets the changes as well. Maybe one could force reload via a server message?

I use the adapter with DAVdroid on my phone, which does so 
automatically, so I never experienced this.

>There is no actual need to keep the ids permanent for evolution, as with every calendar update it accepts the taskwarrior ids; the only thing that's needed though is that within a session, evolution can still access the old/evolution-original id when requesting to know that the changes happened.
>The only problem is that radicale seems to call the adapter every time as opposed to keeping it stable, but maybe there is still a way to store "old ids" in cache for a while and link them to the new entry?

I don't like caching as you can never be sure when to safely delete 
something. Let's hope the etags idea will solve this.

**Comment by mbroedl** [comment]
> Would be great if you could send a pull request to get this integrated.

I realised that I had done that thing I spoke about in the taskwarrior module, so I rewrote it for the adapter; and then I figured, that this probably does not make sense for remind and abook as they only seem to have one file?
With #3  the base-path allows access to all entries regardless of project.

> I fear so as well. Maybe we could have a list in the adapter which projects to export individually.

Thinking about it. If the taskwarrior-module replied with possible pathnames that are not only path/project but also `path/-project` one could reach such thing. It wouldn't look great to list multiple projects, but theoretically it should work. (e.g. subscribing to `path/-project1-project2` and then to `path/project1` and `path/project2` separately).

> I don't like caching as you can never be sure when to safely delete something. Let's hope the etags idea will solve this.

+1

Thanks a lot!

**Comment by jspricke** [comment]
I've pushed some changes that implement etag usage. Note that I changed 
radicale-remind and icstask (and remind and abook as well, in case you 
use that) so you need to get them from git to test. Looking forward to 
your comments.

>I realised that I had done that thing I spoke about in the taskwarrior module, so I rewrote it for the adapter; and then I figured, that this probably does not make sense for remind and abook as they only seem to have one file?

Remind can be split through including other files and providing multiple 
address books is still in my todo list.

>With #3  the base-path allows access to all entries regardless of project.

Thanks! I will try it later today.

>Thinking about it. If the taskwarrior-module replied with possible pathnames that are not only path/project but also `path/-project` one could reach such thing. It wouldn't look great to list multiple projects, but theoretically it should work. (e.g. subscribing to `path/-project1-project2` and then to `path/project1` and `path/project2` separately).

That would be awesome, but we need to announce all lists upfront, so I 
don't think this is feasible.

**Comment by jspricke** [comment]
>I realised that I had done that thing I spoke about in the taskwarrior module, so I rewrote it for the adapter; and then I figured, that this probably does not make sense for remind and abook as they only seem to have one file?

Can you send the version for icstask as well? Thanks!

**Comment by mbroedl** [comment]
Added the change to the icstask and set a pull request. Since I took the basepath instead of adding a project directory to query everything, it was way less changes than originally.

> I've pushed some changes that implement etag usage. Note that I changed radicale-remind and icstask (and remind and abook as well, in case you use that) so you need to get them from git to test. Looking forward to your comments.

Working flawlessly, thanks for that! It still complains when creating a task (see above) but that won't change without caching, so that's fine.

> That would be awesome, but we need to announce all lists upfront, so I don't think this is feasible.

That's a bit problematic indeed. One could maybe create a filterexpression in the config file. Such as: ignore projects with this name, or only taking those with less than... 5 or ten entries. But nothing is really nice.
